![logos](https://files.realpython.com/media/data-cleaning-numpy-pandas.0897550e8675.jpg)

No código de hoje iremos lidar com a Limpeza / Higienização de dados usando <b>Python</b><p>
    Usaremos as biliotecas "mais famosas" a <b>NUMPY</b> e <b>PANDAS</b>

In [1]:
import pandas as pd
import numpy as np

Muitas vezes, você descobrirá que nem todas as categorias de dados em um conjunto de dados são úteis para você. <p>Por exemplo, você pode ter um conjunto de dados contendo informações do aluno (nome, nota, padrão, nomes dos pais e endereço), mas deseja se concentrar na análise das notas dos alunos.

Nesse caso, o endereço ou as categorias de nomes dos pais não são importantes para você. <p>
Manter essas categorias desnecessárias ocupará espaço desnecessário e potencialmente também atrapalhará o tempo de execução.
    Apesar de toda a capacidade de processamento do PYTHON, precisamos ter <b>FOCO</b> 

In [3]:
# df = pd.read_csv e df.head são importantes para o 1º contato com o dataframe
df = pd.read_csv('https://raw.githubusercontent.com/realpython/python-data-cleaning/master/Datasets/BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

No 1º contato nota-se que há muitas colunas que não trazem info's importantes (ou com muitos campos nulos 'NaN')<p>
    Para descartar as mesmas é <b>simples<b> e <b>fácil</b> ...

In [4]:
# ao criar o to_drop é possivel agrupar muitas colunas, claro que você poderia uma a uma também ...
to_drop = ['Edition Statement','Corporate Author','Corporate Contributors','Former owner','Engraver','Contributors','Issuance type','Shelfmarks']
df.drop(to_drop, inplace=True, axis=1)
# É possivel fazer da seguinte forma : df.drop(columns=to_drop, inplace=True)

In [5]:
# veja como o dataframe ficou! O original tinha 15 colunas!
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

Vamos trabalhar com INDEX agora! Estamos trabalhando com mais 8.000 linhas, mas poderiam ser 4.562.300<p>
    Ao indexar o Dataframe, a busca por dados ficará mais ágil

In [7]:
df['Identifier'].is_unique

True

In [8]:
df = df.set_index('Identifier')
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

Caso role a necessidade de acessar um registro especifico, usa-se o ".loc"

In [9]:
# Busca pelo registro '480'
df.loc[480] 

Place of Publication                                               London
Date of Publication                                                  1857
Publisher                                            Wertheim & Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 480, dtype: object

Agora se vc precisar buscar o 5º registro? Simples, utilize o ".iloc"<p>
    Sempre lembrando que a contagem começa no '0' (zero)


In [11]:
df.iloc[4]

Place of Publication                                               London
Date of Publication                                                  1857
Publisher                                            Wertheim & Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 480, dtype: object

Por enquanto eliminamos colunas desnecessárias, criamos índice e até aprendemos como buscar registros<p>
    Perái, agora vem mais coisas!<p>
        Vamos "uniformizar" (nem sei se esse verbo existe risos) o formato dos dados

In [13]:
# como vemos a estrutura de dados do dataframe
df.dtypes

Place of Publication    object
Date of Publication     object
Publisher               object
Title                   object
Author                  object
Flickr URL              object
dtype: object

In [14]:
df.get_dtype_counts()

object    6
dtype: int64

In [17]:
# Adotaremos que o campos de Data de Publicação é importante para os nossos cálculos
df.loc[1905:, 'Date of Publication'].head(5)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
Name: Date of Publication, dtype: object

In [19]:
df.head(20)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London         1879 [1878]   
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   
1280                                   Coventry                1802   
1808                                Christiania                1859   
1905                                    Firenze                1888   
1929                                  Amsterdam         1839, 38-54   
2836                                     Savona                1897   
2854                                     London                1865   
2956                                      Paris             1860-63   
2957                                      Paris                1873   
3017                                Puerto-Rico                1866   
3131                                   New York                1899   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   
1280         Printed by J. Turner   
1808                          NaN   
1905                          NaN   
1929                          NaN   
2836                          NaN   
2854               E. Moxon & Co.   
2956                          NaN   
2957                          NaN   
3017                          NaN   
3131                    W. Abbatt   

                                                        Title  \
Identifier                                                      
206                         Walter Forbes. [A novel.] By A. A   
216         All for Greed. [A novel. The dedication signed...   
218         Love the Avenger. By the author of “All for Gr...   
472         Welsh Sketches, chiefly ecclesiastical, to the...   
480         [The World in which I live, and my place in it...   
481         [The World in which I live, and my place in it...   
519         Lagonells. By the author of Darmayne (F. E. A....   
667         The Coming of Spring, and other poems. By J. A...   
874         A Warning to the inhabitants of England, and L...   
1143        A Satyr against Vertue. (A poem: supposed to b...   
1280        An Account of the many and great Loans, Benefa...   
1808        Erindringer som Bidrag til Norges Historie fra...   
1905        Gli Studi storici in terra d'Otranto ... Framm...   
1929        De Aardbol. Magazijn van hedendaagsche land- e...   
2836        Cronache Savonesi dal 1500 al 1570 ... Accresc...   
2854        See-Saw; a novel ... Edited [or rather, writte...   
2956        Géodésie d'une partie de la Haute Éthiopie,...   
2957                                      [With eleven maps.]   
3017        [Historia geográfica, civil y politica de la ...   
3131        The Crisis of the Revolution, being the story ...   

                                                       Author  \
Identifier                                                      
206                                                     A. A.   
216                    

Com o foco ainda no campo de 'Data de Publicação', vamos deixa-la com o mesmo formato! Como? <p>
    Vamos tratar conteúdos como por exemplo: '1879 [1878]', 'NaN' e '1839, 38-54'

In [21]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

Alterar o foramto do campo é a próxima missão!

In [26]:
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtypes

dtype('float64')

MISSÃO DADA É MISSÃO COMPRIDA !<p>
    Dessa forma conseguimos usar tam campo em cálculos

In [27]:
df['Date of Publication'].isnull().sum() / len(df)

0.11717147339205986